pkg/deploy/controller/generictrigger/factory.go

```diff
@@ -89,24 +92,64 @@ func (c *DeploymentTriggerController) waitForSyncedStore(ready chan<- struct{},
 
 func (c *DeploymentTriggerController) addDeploymentConfig(obj interface{}) {
 	dc := obj.(*deployapi.DeploymentConfig)
+
+	// No need to enqueue deployment configs that have no triggers or are paused.
 	if len(dc.Spec.Triggers) == 0 || dc.Spec.Paused {
 		return
 	}
+	// We don't want to compete with the main deployment config controller. Let's process this
+	// config once it's synced.
+	if !deployutil.HasSynced(dc, dc.Generation) {
+		return
+	}
+
 	c.enqueueDeploymentConfig(dc)
 }
 
 func (c *DeploymentTriggerController) updateDeploymentConfig(old, cur interface{}) {
-	// If there is no generation update for this deployment config,
-	// we have a good indication to not enqueue it.
 	newDc := cur.(*deployapi.DeploymentConfig)
 	oldDc := old.(*deployapi.DeploymentConfig)
-	if newDc.Generation == oldDc.Generation {
+
+	// A periodic relist will send update events for all known deployment configs.
+	if newDc.ResourceVersion == oldDc.ResourceVersion {
 		return
 	}
-
+	// No need to enqueue deployment configs that have no triggers or are paused.
 	if len(newDc.Spec.Triggers) == 0 || newDc.Spec.Paused {
 		return
 	}
+	// We don't want to compete with the main deployment config controller. Let's process this
+	// config once it's synced. Note that this does not eliminate conflicts between the two
+	// controllers because the main controller is constantly updating deployment configs as
+	// owning replication controllers and pods are updated.
+	if !deployutil.HasSynced(newDc, newDc.Generation) {
+		return
+	}
+	// Enqueue the deployment config if it hasn't been deployed yet.
+	if newDc.Status.LatestVersion == 0 {
+		c.enqueueDeploymentConfig(newDc)
+		return
+	}
+	// Compare deployment config templates before enqueueing. This reduces the amount of times
+	// we will try to instantiate a deployment config at the expense of duplicating some of the
+	// work that the instantiate endpoint is already doing but I think this is fine.
+	shouldInstantiate := true
+	latestRc, err := c.rcLister.ReplicationControllers(newDc.Namespace).Get(deployutil.LatestDeploymentNameForConfig(newDc))
+	if err != nil {
+		// If we get an error here it may be due to the rc cache lagging behind. In such a case
+		// just defer to the api server (instantiate REST) where we will retry this.
+		glog.V(2).Infof("Cannot get latest rc for dc %s:%d (%v) - will defer to instantiate", deployutil.LabelForDeploymentConfig(newDc), newDc.Status.LatestVersion, err)
+	} else {
+		initial, err := deployutil.DecodeDeploymentConfig(latestRc, c.codec)
+		if err != nil {
+			glog.V(2).Infof("Cannot decode dc from replication controller %s: %v", deployutil.LabelForDeployment(latestRc), err)
+			return
+		}
+		shouldInstantiate = !reflect.DeepEqual(newDc.Spec.Template, initial.Spec.Template)
```

Comment:
As expensive as a deep equal check can be? We were running this in the trigger controller already prior to refactoring the controller to use the instantiate endpoint.